### PR TITLE
Vehicle Functions: Update daily

### PR DESCRIPTION
**Pull request by duncanspumpkin**
There is some horrible branches in this code. I can't see an easy way to rework it to not be as horrible whilst still matching vanilla logic.

### CODE REVIEW

File: src/OpenLoco/src/Vehicles/VehicleHead.cpp [hunk 1]
```diff
@@ -179,12 +182,158 @@ namespace OpenLoco::Vehicles
         calculateRefundCost();
     }
 
+    // 0x004B8340
+    static void RecalculateTrainMinReliability(VehicleHead& head)
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Vehicles/VehicleHead.cpp:186]: Minor code style violation:
```suggestion
    static void recalculateTrainMinReliability(VehicleHead& head)
```

File: src/OpenLoco/src/Vehicles/VehicleHead.cpp [hunk 2]
```diff
@@ -179,12 +182,158 @@ namespace OpenLoco::Vehicles
...
+        const auto getReliabilty = [](const Car& car) -> uint16_t {
+            if (car.front->reliability == 0)
+            {
+                return 0xFFFFU;
+            }
+            return car.front->reliability;
+        };
+        const auto minReliability = (*std::min_element(train.cars.begin(), train.cars.end(), [&getReliabilty](const Car& carLhs, const Car& carRhs) {
+                                        return getReliabilty(carLhs) < getReliabilty(carRhs);
+                                    })).front->reliability;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Vehicles/VehicleHead.cpp:198]: Can we improve the formatting for this, somehow?

File: src/OpenLoco/src/Vehicles/VehicleHead.cpp [hunk 3]
```diff
@@ -179,12 +182,158 @@ namespace OpenLoco::Vehicles
...
+        bool resetStoppedTimeout = true;
+        if (status == Status::stopped)
+        {
+            if (mode == TransportMode::road)
+            {
+                if (tileX != -1
+                    && hasVehicleFlags(VehicleFlags::commandStop))
+                {
+                    if (Tutorial::state() == Tutorial::State::none)
+                    {
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Vehicles/VehicleHead.cpp:214]: Could we simplify these `if` blocks a little to reduce nesting? All `else` branches just set `resetStoppedTimeout = false`, after all.

**Review by AaronVanGeffen — COMMENTED** [review]
Comments with respect to code style aside, have you run any replay tests for this yet?

File: src/OpenLoco/src/Vehicles/VehicleHead.cpp (hunk 3 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/Vehicles/VehicleHead.cpp:214]: I'm not sure how to do that and keep the same results. Not everything has an else block as they don't all reset the stopped timeout. If it was as simple as all have an else then yes i could see how to do it.

Comment (Lev135) [reply src/OpenLoco/src/Vehicles/VehicleHead.cpp:214]: It seems like it is a bit better to use `resetStoppedTimeout = false` as initial value both logically (by default we don't need to do anything) and for readability (we don't need 1 else in this case):
```c++
bool resetStoppedTimeout = false;
if (status == Status::stopped)
{
    if (mode == TransportMode::road)
    {
        if (tileX != -1
            && hasVehicleFlags(VehicleFlags::commandStop))
        {
            if (Tutorial::state() == Tutorial::State::none)
            {
                var_79++;
                if (var_79 >= 20)
                {
                    GameCommands::VehicleChangeRunningModeArgs args{};
                    args.head = head;
                    args.mode = GameCommands::VehicleChangeRunningModeArgs::Mode::startVehicle;
                    auto regs = static_cast<Interop::registers>(args);
                    regs.bl = GameCommands::Flags::apply;
                    GameCommands::vehicleChangeRunningMode(regs);
                    resetStoppedTimeout = true;
                }
            }
        }
        else
        {
            resetStoppedTimeout = true;
        }
    }
}
else
{
    resetStoppedTimeout = true;
}
```
[The change process](https://gist.github.com/Lev135/cc06a189ced5ce19b3362dae9a30c48c) was very simple, so I hope I don't broke anything. But, please, check me too

**Comment by AaronVanGeffen** [comment]
What's still needed for this PR?

File: src/OpenLoco/src/Vehicles/VehicleHead.cpp (hunk 2 above)
Comment (LeftofZen) [reply src/OpenLoco/src/Vehicles/VehicleHead.cpp:198]: just refactoring the predicate out makes it fairly nice

```const auto pred = [&getReliabilty](const Car& carLhs, const Car& carRhs) { return getReliabilty(carLhs) < getReliabilty(carRhs); };
const auto minReliability = (*std::min_element(train.cars.begin(), train.cars.end(), pred)).front->reliability;
```

File: src/OpenLoco/src/Vehicles/VehicleHead.cpp [hunk 4]
```diff
@@ -179,12 +182,158 @@ namespace OpenLoco::Vehicles
         calculateRefundCost();
     }
 
+    // 0x004B8340
+    static void RecalculateTrainMinReliability(VehicleHead& head)
+    {
+        Vehicle train(head);
+        const auto getReliabilty = [](const Car& car) -> uint16_t {
```
Comment (LeftofZen) [inline src/OpenLoco/src/Vehicles/VehicleHead.cpp:189]: typo. `getReliability`

File: src/OpenLoco/src/Vehicles/VehicleHead.cpp [hunk 5]
```diff
@@ -179,12 +182,158 @@ namespace OpenLoco::Vehicles
...
+        {
+            if (mode == TransportMode::road)
+            {
+                if (tileX != -1
+                    && hasVehicleFlags(VehicleFlags::commandStop))
+                {
+                    if (Tutorial::state() == Tutorial::State::none)
+                    {
+                        var_79++;
+                        if (var_79 >= 20)
```
Comment (LeftofZen) [inline src/OpenLoco/src/Vehicles/VehicleHead.cpp:216]: var we name `var_79` now

File: src/OpenLoco/src/World/CompanyAi.cpp [hunk 6]
```diff
@@ -963,4 +963,21 @@ namespace OpenLoco
             GameCommands::doCommand(args, GameCommands::Flags::apply);
         }
     }
+
+    void removeEntityFromThought(AiThought& thought, EntityId id)
+    {
+        auto iter = std::find(std::begin(thought.var_66), std::end(thought.var_66), id);
+        if (iter == std::end(thought.var_66))
```
Comment (LeftofZen) [inline src/OpenLoco/src/World/CompanyAi.cpp:970]: can we name `var_66`?

File: src/OpenLoco/src/World/CompanyAi.cpp [hunk 7]
```diff
@@ -963,4 +963,21 @@ namespace OpenLoco
...
+        {
+            return;
+        }
+        // Original would copy the value from var_66 + 2 which
+        // would mean it would copy currency var_76 if var_44 was 7
+        // I don't think that is possible but lets just add an assert.
+        assert(thought.var_44 < 7);
```
Comment (LeftofZen) [inline src/OpenLoco/src/World/CompanyAi.cpp:977]: similarly can we name `var_44`?

File: src/OpenLoco/src/Vehicles/VehicleHead.cpp [hunk 8]
```diff
@@ -179,12 +182,158 @@ namespace OpenLoco::Vehicles
...
+                {
+                    front.var_68--;
+                }
+                if (front.var_68 != 0)
+                {
+                    sub_4BA873(front);
+                    front.breakdownFlags |= BreakdownFlags::breakdownPending;
+                }
+            }
+            auto newReliabilty = front.reliability;
```
Comment (LeftofZen) [inline src/OpenLoco/src/Vehicles/VehicleHead.cpp:330]: typo `newReliability`

File: src/OpenLoco/src/Vehicles/VehicleHead.cpp (hunk 3 above)
Comment (LeftofZen) [reply src/OpenLoco/src/Vehicles/VehicleHead.cpp:214]: You can trvially combine the inner-most `if` if you inline the `var_79++`:

```
 if (status == Status::stopped)
 {
     if (mode == TransportMode::road)
     {
         if (tileX != -1 && hasVehicleFlags(VehicleFlags::commandStop))
         {
             if (Tutorial::state() == Tutorial::State::none && ++var_79 >= 20)
             {
                 GameCommands::VehicleChangeRunningModeArgs args{};
                 args.head = head;
                 args.mode = GameCommands::VehicleChangeRunningModeArgs::Mode::startVehicle;
                 auto regs = static_cast<Interop::registers>(args);
                 regs.bl = GameCommands::Flags::apply;
                 GameCommands::vehicleChangeRunningMode(regs);
             }
             else
             {
                 resetStoppedTimeout = false;
             }
         }
     }
     else
     {
         resetStoppedTimeout = false;
     }
 }
```

File: src/OpenLoco/src/Vehicles/VehicleHead.cpp [hunk 9]
```diff
@@ -179,12 +182,158 @@ namespace OpenLoco::Vehicles
...
+        {
+            var_79 = 0;
+        }
+
+        if (status == Status::crashed || status == Status::stuck)
+        {
+            crashedTimeout = Math::Bound::add(crashedTimeout, 1U);
+            if (!CompanyManager::isPlayerCompany(owner))
+            {
+                if (crashedTimeout > 14 && var_60 != 0xFF)
```
Comment (LeftofZen) [inline src/OpenLoco/src/Vehicles/VehicleHead.cpp:251]: can we add a constant for the `14`, and can we rename `var_60`?

File: src/OpenLoco/src/Vehicles/VehicleHead.cpp [hunk 10]
```diff
@@ -179,12 +182,158 @@ namespace OpenLoco::Vehicles
...
+        {
+            auto& front = *car.front;
+            if (front.reliability == 0)
+            {
+                continue;
+            }
+
+            if (front.hasBreakdownFlags(BreakdownFlags::brokenDown))
+            {
+                front.var_6A--;
```
Comment (LeftofZen) [inline src/OpenLoco/src/Vehicles/VehicleHead.cpp:311]: name `var_6A` and `var_6B``

**Review by LeftofZen — CHANGES_REQUESTED** [review]
Multiple minor code changes requested. Branch also has merge conflicts

**Comment by duncanspumpkin** [comment]
There is some sort of divergence in this. I've narrowed it down to the loop at L312 or the minReliabilityCount but i can't tell what ive done wrong. I didn't implement it exactly as per the assembly as it looked like there were some easy simplifications.

File: src/OpenLoco/src/Vehicles/VehicleBogie.cpp [hunk 11]
```diff
@@ -312,7 +312,7 @@ namespace OpenLoco::Vehicles
             reliabilityFactor /= 16;
 
             auto& prng = gPrng1();
-            int32_t randVal = (prng.randNext(65535) * reliabilityFactor / 2) / 65536;
+            int32_t randVal = (prng.randNext(65535) * (reliabilityFactor / 2)) / 65536;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Vehicles/VehicleBogie.cpp:315]: this fix causes replays to diverge but i checked and it is how it is meant to work so will record new masters.

File: src/OpenLoco/src/World/CompanyAi.cpp (hunk 6 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/World/CompanyAi.cpp:1432]: I'm still not sure on what to name it

File: src/OpenLoco/src/World/CompanyAi.cpp (hunk 7 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/World/CompanyAi.cpp:1439]: I've looked at var_44 and var_66 again and I still don't know enough to name them. I think best to leave it for now.

File: src/OpenLoco/src/Vehicles/VehicleHead.cpp (hunk 9 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/Vehicles/VehicleHead.cpp]: I've added a constant but again don't know enough about var_60 atm.

File: src/OpenLoco/src/GameCommands/Vehicles/CreateVehicle.cpp [hunk 12]
```diff
@@ -450,7 +431,7 @@ namespace OpenLoco::GameCommands
         newHead->var_61 = 0xFFU;
         newHead->totalRefundCost = 0;
         newHead->lastAverageSpeed = 0_mph;
-        newHead->var_79 = 0;
+        newHead->restartStoppedCarsTimeout = 0;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/GameCommands/Vehicles/CreateVehicle.cpp:434]: Nice to have this identified!

File: src/OpenLoco/src/Vehicles/Vehicle.h [hunk 13]
```diff
@@ -646,10 +646,10 @@ namespace OpenLoco::Vehicles
...
         uint8_t var_60;
         uint8_t var_61;
-        uint32_t refundCost;  // 0x62 front bogies only
-        uint16_t reliability; // 0x66 front bogies only
-        uint16_t var_68;
-        uint8_t var_6A;
+        uint32_t refundCost;         // 0x62 front bogies only
+        uint16_t reliability;        // 0x66 front bogies only
+        uint16_t timeoutToBreakdown; // 0x68 front bogies only (days) counts down to the next breakdown 0xFFFFU disables this
+        uint8_t breakdownTimeout;    // 0x6A front bogies only (days)
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Vehicles/Vehicle.h:652]: Nice to have these identified!

**Review by AaronVanGeffen — APPROVED** [review]
Changes look good to me! You mentioned a divergence in replays due to an RNG fix. Will you be re-recording replays as of this PR, then?

**Review by LeftofZen — APPROVED** [review]
seems fine in playtesting; my vehicles still broke down

**Comment by duncanspumpkin** [comment]
[MMTest100000_base.zip](https://github.com/user-attachments/files/16740461/MMTest100000_base.zip)
Update replay ref for refererence.

**Pull request merged**